src/include/platform/Mdns.h

```diff
@@ -0,0 +1,168 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#ifndef CHIP_PLATFORM_MDNS_H_
+#define CHIP_PLATFORM_MDNS_H_
+
+#include <stdint.h>
+
+#include "core/CHIPError.h"
+#include "inet/IPAddress.h"
+#include "inet/InetInterface.h"
+
+namespace chip {
+namespace DeviceLayer {
+
+static constexpr uint8_t kMdnsNameMaxSize  = 32;
+static constexpr uint8_t kMdnsTypeMaxSize  = 32;
+static constexpr uint16_t kMdnsTextMaxSize = 64;
+
+enum class MdnsServiceProtocol : uint8_t
+{
+    kMdnsProtocolUdp = 0,
+    kMdnsProtocolTcp,
+    kMdnsProtocolUnknown = 255,
+};
+
+struct TextEntry
+{
+    const uint8_t * mData;
+    size_t mSize;
+};
+
+struct MdnsService
+{
+    char mName[kMdnsNameMaxSize + 1];
+    char mType[kMdnsTypeMaxSize + 1];
+    MdnsServiceProtocol mProtocol;
+    uint16_t mPort;
+    chip::Inet::InterfaceId interface;
+    TextEntry * mTextEntryies;
+    size_t mTextEntrySize;
+};
+
+struct MdnsResolveResult
+{
+    MdnsService mService;
+    chip::Inet::IPAddress mAddress;
+};
+
+/**
+ * The callback function for mDNS resolve.
+ *
+ * The callback function SHALL NOT take the ownership of the result->mService.mTextEntries
+ * memory.
+ *
+ * @param[in] context     The context passed to ChipMdnsBrowse or ChipMdnsResolve.
+ * @param[in] result      The mdns resolve result, can be nullptr if error happens.
+ * @param[in] error       The error code.
+ *
+ */
+using MdnsResolveCallback = void (*)(void * context, MdnsResolveResult * result, CHIP_ERROR error);
+
+/**
+ * The callback function for mDNS browse.
+ *
+ * The callback function SHALL NOT take the ownership of the service->mTextEntries
```

Comment:
how many times does the BrowseCallback fire?

Reply:
It will be called only once for each browse when the query stage ends.